### PR TITLE
Support for GPU NUFFT

### PR DESCRIPTION
**Pull request by chaithyagr**
This PR Holds the support for GPU based NUFFT which is critical for 3D runs.

### CODE REVIEW

**Comment by chaithyagr** [comment]
The drop in coverage is as we added a local test for GPU which we cant test on Travis, but we still to test locally whenever needed

File: mri/reconstruct/fourier.py [hunk 1]
```diff
@@ -208,3 +215,227 @@ def adj_op(self, x):
...
+
+            self.nufftObj.plan(om=self.samples,
+                               Nd=self.shape,
+                               Kd=self.Kd,
+                               Jd=self.Jd,
+                               batch=1,  # self.nb_coils,
+                               ft_axes=tuple(range(samples.shape[1])),
+                               radix=None)
+
+        elif self.platform == 'gpu':
```
Comment (LElgueddari) [inline mri/reconstruct/fourier.py]: You should consider excluding the GPU version from coverall. Check out [this](https://coverage.readthedocs.io/en/coverage-4.2/excluding.html) documentation.

Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: As discussed, we decided to not do this as the goal is to not have 100 percent coverage but to know what parts arent tested. As these codes will be constantly used going forward, we plan to keep this ON. Maybe in future we could start running even the GPU versions of codes which could bump up the Coverage numbers

**Comment by chaithyagr** [comment]
Rebased the codes, moved the code base of NFFT as NUFFT with platform as 'cpu'
Also, I have modified the tags as 'cuda' and 'opencl' than 'multi-cpu' and 'gpu' as it makes more sense.

Added 2 more local tests for NUFFT with opencl and they are working fine. 
Adding for re-review and merge.

**Comment by chaithyagr** [comment]
@LElgueddari , I had a discussion with @zaccharieramzi  on the way we would want to structure the classes for NFFT. Here is what we felt is best implementation:

--- Fourier Base -> Ensures that the operator and adjoint Operator is implemented. 
        
--- NFFT(derived from Fourier_base) -> This is a base class for carrying out NFFT. This will hold some basic stuff which  is used by the implementations 
  -----> In this class, we will have a variable (say 'calculator') , which would hold an object of class which carries out implementation in CPU or on GPU with CUDA or OpenCL. **Note that this class will not be a parent class for NFFT.** We will initialize the right variable and do the same implementation based on what implementation we prefer.

Does this sound good?

I agree that this is cleanest implementation possible that follows the OOP concepts the right way.

**Comment by chaithyagr** [comment]
Updated codes @LElgueddari , you can re-review and suggest any more changes. We will rebase this commit once #41 is at mainline

File: mri/numerics/fourier.py [hunk 2]
```diff
@@ -13,5 +13,5 @@
 
 # Package import
 from mri.reconstruct.fourier import FFT2
-from mri.reconstruct.fourier import NFFT
+from mri.reconstruct.fourier import NUFFT
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/numerics/fourier.py]: Yes, on it

File: mri/reconstruct/fourier.py [hunk 3]
```diff
@@ -26,6 +27,12 @@
     warnings.warn("pynfft python package has not been found. If needed use "
                   "the master release.")
     pass
+try:
+    from pynufft import NUFFT_hsa, NUFFT_cpu
+except Exception:
+    warnings.warn("pynufft python package has not been found. If needed use "
+                  "the master release. Till then you cant use NUFFT on GPU")
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Both are same in my opinion, changing to cannot now..

File: mri/reconstruct/fourier.py [hunk 4]
```diff
@@ -208,3 +215,273 @@ def adj_op(self, x):
...
+        Kd: int or tuple
+            int or tuple indicating the size of the frequency grid,
+            for regridding. If int, will be evaluated
+            to (Kd,)*nb_dim of the image
+        Jd: int or tuple
+            Size of the interpolator kernel. If int, will be evaluated
+            to (Jd,)*dims image
+        n_coils: int
+            Number of coils used to acquire the signal in case of multiarray
+            receiver coils
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Done

File: mri/reconstruct/fourier.py [hunk 5]
```diff
@@ -208,3 +215,273 @@ def adj_op(self, x):
...
+            self.nufftObj = NUFFT_hsa(API='ocl',
+                                      platform_number=None,
+                                      device_number=None,
+                                      verbosity=verbosity)
+
+            self.nufftObj.plan(om=self.samples,
+                               Nd=self.shape,
+                               Kd=self.Kd,
+                               Jd=self.Jd,
+                               batch=1,  # self.nb_coils,
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Yes thats the plan, will add TODO

**Comment by chaithyagr** [comment]
Updated the codes and also rebased them so that the examples work fine. @LElgueddari can you please fix the build issue and then merge this PR so that we are sure that nothing is breaking

File: examples/non_cartesian_reconstruction.py [hunk 6]
```diff
@@ -49,7 +49,7 @@
 # We then reconstruct the zero order solution as a baseline
 
 # Get the locations of the kspace samples and the associated observations
-fourier_op = NFFT(samples=kspace_loc, shape=image.shape)
+fourier_op = NonCartesianFFT(samples=kspace_loc, shape=image.shape)
```
Comment (LElgueddari) [inline examples/non_cartesian_reconstruction.py]: add the implementation parameter even if it is the default value

File: mri/parallel_mri/extract_sensitivity_maps.py [hunk 7]
```diff
@@ -161,7 +161,7 @@ def get_Smaps(k_space, img_shape, samples, thresh,
         for l in range(Smaps_shape[2]):
             Smaps[l] = pfft.ifftshift(pfft.ifft2(pfft.fftshift(k_space[l])))
     elif mode == 'NFFT':
-        fourier_op = NFFT(samples=samples, shape=img_shape)
+        fourier_op = NonCartesianFFT(samples=samples, shape=img_shape)
```
Comment (LElgueddari) [inline mri/parallel_mri/extract_sensitivity_maps.py]: add `implementation` parameters

File: mri/reconstruct/fourier.py [hunk 8]
```diff
@@ -207,5 +231,284 @@ def adj_op(self, x):
...
+        if implementation == 'cpu':
+            self.implementation = NFFT(samples=samples, shape=shape,
+                                       n_coils=self.nb_coils)
+        elif implementation == 'cuda' or implementation == 'opencl':
+            self.implementation = NUFFT(samples=samples, shape=shape,
+                                        platform=implementation,
+                                        n_coils=self.nb_coils)
+        else:
+            raise ValueError('Bad implementation ' + implementation +
+                             ' chosen')
```
Comment (LElgueddari) [inline mri/reconstruct/fourier.py]: I'm not sure that this is a good message error, I would recommend to add the different options inside message

File: mri/test/test_fourier_adjoint.py [hunk 9]
```diff
@@ -26,13 +27,14 @@ def setUp(self):
         """
         self.N = 64
         self.max_iter = 10
+        self.num_channels = [1, 16]
```
Comment (LElgueddari) [inline mri/test/test_fourier_adjoint.py]: For the test `num_channels = [1, 16]` is probably too much, instead, you could reduce it to `[1, 2]`

**Comment by chaithyagr** [comment]
> FFT2 should also be changed to take into account the number of coils to be more consistent with the rest of the code.
> FFT2 should be changed to FFTN to avoid duplicate code for he 3D case

Can we get this done when we are changing FFT2? ie in #45 ? I feel that is best place for this

File: mri/reconstruct/fourier.py [hunk 10]
```diff
@@ -189,8 +204,17 @@ def op(self, img):
...
-        return np.copy(self.plan.trafo()) / np.sqrt(self.plan.M)
+        if self.nb_coils == 1:
+            coeff = self._op(img)
+        else:
+            coeff = [self._op(img[i])
+                     for i in range(self.nb_coils)]
+            coeff = np.asarray(coeff)
+        return coeff
+
+    def _adj_op(self, x):
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Duplicate, resolving here. We will discuss this above.

File: mri/reconstruct/fourier.py [hunk 11]
```diff
@@ -206,5 +230,284 @@ def adj_op(self, x):
         img: np.ndarray
             inverse 2D discrete Fourier transform of the input coefficients.
         """
-        self.plan.f = x
-        return np.copy(self.plan.adjoint()) / np.sqrt(self.plan.M)
+        if self.nb_coils == 1:
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Ack.

File: mri/reconstruct/fourier.py [hunk 12]
```diff
@@ -206,5 +230,284 @@ def adj_op(self, x):
...
+        elif Jd is None:
+            # Preferential option
+            self.Jd = (5,) * self.dim
+
+        for (i, s) in enumerate(shape):
+            assert (self.shape[i] <= self.Kd[i]), 'size of frequency grid' + \
+                                                  'must be greater or equal ' \
+                                                  'than the image size'
+
+        print('Creating the NUFFT object...')
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Yes Will print only if verbose=1

File: examples/non_cartesian_reconstruction.py (hunk 6 above)
Comment (chaithyagr) [reply examples/non_cartesian_reconstruction.py]: Sure

File: mri/reconstruct/fourier.py (hunk 8 above)
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Yes, will add options

File: mri/test/test_fourier_adjoint.py (hunk 9 above)
Comment (chaithyagr) [reply mri/test/test_fourier_adjoint.py]: Sure.

File: mri/reconstruct/fourier.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: I dont think I quite understand what needs to be added here.

File: mri/reconstruct/fourier.py [hunk 13]
```diff
@@ -189,8 +204,17 @@ def op(self, img):
         x: np.ndarray
             masked Fourier transform of the input image.
         """
-        self.plan.f_hat = img
-        return np.copy(self.plan.trafo()) / np.sqrt(self.plan.M)
+        if self.nb_coils == 1:
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Same as below. Am I missing something here?

File: mri/reconstruct/fourier.py [hunk 14]
```diff
@@ -170,10 +179,16 @@ def __init__(self, samples, shape):
...
             self.samples = normalize_frequency_locations(self.samples)
+        # TODO Parallelize this if possible
+        self.nb_coils = n_coils
         self.plan = pynfft.NFFT(N=shape, M=len(samples))
-        self.shape = shape
         self.plan.x = self.samples
         self.plan.precompute()
+        self.shape = shape
+
+    def _op(self, img):
```
[the remark this answers is not in the document]
Comment (chaithyagr) [reply mri/reconstruct/fourier.py]: Resolving for now, please reopen for discussion.

**Comment by chaithyagr** [comment]
I made a few modifications for travis to ensure a passing build. @LElgueddari can you please merge?

**Pull request merged**